File: Admin/lib/models/order.dart
//
//
//
import 'package:admin/models/cart.dart';
import 'package:admin/models/customer.dart';

class Order {
  ///
  int? id;
  int? vendorId;
  int? customerId;
  String? status;
  int? cartId;
  Customer? customer;
  Cart? cart;

  int? deliveryCost;
  int? total;
  int? deliveryManId;
  String? pickupAddress;
  String? deliveryAddress;
  String? notes;

  ///TODO : Add delivered at (Date)

  ///
  Order({
    this.id,
    this.vendorId,
    this.status,
    this.customerId,
    this.cartId,
    this.customer,
    this.cart,
    this.total,
    this.deliveryAddress,
    this.pickupAddress,
    this.deliveryManId,
    this.notes,
    this.deliveryCost,
  });

  ///
  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: json['id'],
      cartId: json['cart_id'],
      vendorId: json['vendor_id'],
      customerId: json['customer_id'],
      deliveryManId: json['deliveryman_id'],
      total: json['total'] != null ? (json['total'] as int) : null,
      status: json['status'],
      deliveryAddress: json['delivery_address'],
      pickupAddress: json['pickup_address'],
      notes: json['notes'],
      deliveryCost: json['delivery_cost'],
      customer: json['customer'] != null
          ? Customer.fromJson(json['customer'])
          : null,
      cart: json['cart'] != null ? Cart.fromJson(json['cart']) : null,
    );
  }

  ///
  Map<String, dynamic> toJson() => {
    'id': id,
    'vendor_id': vendorId,
    'customer_id': customerId,
    'deliveryman_id': deliveryManId,
    'total': total,
    'status': status,
    'pickup_address': pickupAddress,
    'delivery_address': deliveryAddress,
    'notes': notes,
    'delivery_cost': deliveryCost,
    'cart_id': cartId,
    'customer': customer!.toJson(),
    'cart': cart!.toJson(),
  };
}

File: absher-api/app/Models/Order.php
<?php

namespace App\Models;

use App\OrderStatus;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Order extends Model
{
    ///
    use HasFactory;

    ///
    protected $fillable = [
        'customer_id',
        'vendor_id',
        'deliveryman_id',
        'pickup_address',
        'delivery_address',
        'status',
        'scheduled_at',
        'delivered_at',
        'notes',
        'total',
        'cart_id',
        'delivery_cost'
    ];


    ///
    protected $casts = [
        'scheduled_at' => 'datetime',
        'delivered_at' => 'datetime',
        // 'status' => OrderStatus::class
    ];

    ///
    public function cart()
    {
        return $this->belongsTo(Cart::class);
    }

    ///
    public function customer()
    {
        return $this->belongsTo(User::class, 'customer_id');
    }

    ///
    public function vendor()
    {
        return $this->belongsTo(User::class, 'vendor_id');
    }

    ///
    public function deliveryman()
    {
        return $this->belongsTo(User::class, 'deliveryman_id');
    }

    //
    public function deliveryRequests()
    {
        return $this->hasMany(DeliveryRequest::class);
    }

    public function approvedDeliveryRequest()
    {
        return $this->hasOne(DeliveryRequest::class)->where('status', 'approved');
    }
}
